### ISSUE TITLE
Update code for version 16.00.47

### ISSUE THREAD

**Issue by smalers** [OpenCDSS/cdss-app-statemod-fortran#34]
Ray Bennett has provided StateMod 16.00.47 code in a zip file named `StateMod_160047.zip` available on the State's Google Drive.  The zip file contains all the source code, but I don't see a `makefile`.  These files need to be committed to the repository once they are confirmed to compile.

I recently updated the development environment to use MSys2 64-bit with MinGW 64-bit and `gfortran` 10.2.0.  The development environment needs to be tested out and executable compiled. Once that is done, can move on to testing as part of OpenCDSS2 project.

**Comment by Rayrbennett** [comment]
Sorry, do you need a make file?  (or you can copy the one I provided in my email.)
Ray

**Comment by smalers** [comment]
I am going to use this issue to record some notes and then will pull into developer documentation.

Ray provided the 16.00.47 update in a zip file available on the State's Google drive.

1. I tagged the starting point before doing any work, as `StateMod-15.0014-before-16`.
2. I did a file listing of existing and zip file contents using `ls -1`.
   1. Existing code has 258 `*.for` and 2 `*.inc` files.  New code has 264 `*.for` and 2 `*.inc` files.  If no old files have been renamed or deleted, this means that 6 new files have been added.  Because `license.for` is not included in Ray's version, 7 new files are expected (see below).  I'm going to need to re-add logic to the main program to print the license.
   2. Existing code has `AM.bat`, `AUTOMAKE.RSP`, `README.md`, `automake.fig`, `license.for`, and `makefile`, which are not in the new files.
   3. New files have `FlowRes.for`, `JMFlow.for`, `JMStore.for`, `chekav3.for`, `divmultR.for`, `outJM.for`, `outWW.for`,
3. I used KDiff3 to compare individual file differences for files that exist in both versions.
   1. None of Ray's files have the GPL 3 notice at the top.  I am going to have to re-add to all of the files so that kdiff3 shows the same content.  I expect that many files will have this as the only difference so that once the notice is added there will be many fewer files with differences.  **Need to make sure that any future work starts with the updated version so don't have to do that again.**
   2. A number of typos exist that were fixed before, for example `mininum` in `checkeva.for`.  I will fix this so that kdiff3 shows the same content (unless there are changes in the new files).

**Comment by smalers** [comment]
I accidentally closed the issue. Reopen since work in progress.

**Comment by Rayrbennett** [comment]
6 new files look correct.  
Expect lots of changes throughout. Some changes are edits, most are clean up on comments.  Suspect the tab issue will persist; sorry I never got that fully implemented with my editor. 

Ray

**Comment by smalers** [comment]
I have to inspect every file.  My approach is:

1. Use kdiff3 to visually compare every file and use its "Next difference" button.
2. If the difference is only the GPL 3 notice at top, keep the old file, change the copyright to have end year of 2021.
3. If there are minor changes, make the files agree.
4. If there are more changes, copy Ray's file and re-insert the GPL notice.
5. Also fix:
    1. I considered replacing tabs with spaces, usually in comments, or extra at ends of lines, sometimes in code (fix open and closing statements to align.  However, I quickly grew weary of fixing tab/alignment problems because I don't have time, especially when the code otherwise did not change.  A lot more could be done to replace tabs with space-2 and make sure alignment is OK.
    2. Fix typos, especially common ones like `Initilize`. `calcuate` if obvious.  There are probably a lot more typos.
    3. Remove "last change" comments that are obviously wrong, when I notice.  Usually there are more recent history notes.

**Comment by smalers** [comment]
Ray Bennett provided a zip file with updated code.  There are over 50 files with differences based on kdiff3 comparison, some of which are log files and other files that don't need to be put in the repository.  Below is a list of each modified file that is in the repository.  After inspecting, I copied Ray's files into the working files to test compiling.  I tried to visually confirm that the label numbers are correct for loop.  I did notice a few other things, which I've noted in bold in the table below.

| **File** | **Comment** |
| -- | -- |
| `AM.bat` | Have in the repo but Ray has not supplied recent version. |
| `AUTOMAKE.DEP` | Binary file - use gitignore. |
| `AUTOMAKE.RSP` | Automatically created by Lahey build tools from list of files in folder.  **This file seems to include `_gfortran` and `_linux` files that should not be included for Lahey because it will result in duplicate routine names.  Ray needs to remove the files from the folder or rename with extensions that are not handled as code.  Then they won't get included by Lahey build tools. |
| `AUTOMAKE2.RSP` | Maybe a manual backup?  Not including in the repository. |
| `automake_dep.old` and `automake_dep_2020-12-07.od` | Backups.  Not including in the repository. |
| `bomsec.for` | Added do loop end `continue`.  |
| `carrpl.for` | Added do loop end `continue`. |
| `datinp.for` |  Added do loop end `continue`. |
| `demcons.for` |  Added do loop end `continue`. |
| `divcar2.for` |  Added do loop end `continue`. |
| `divimp2.for` |  Added do loop end `continue`. |
| `divres.for` |  Added do loop end `continue`. |
| `divrpl.for` |  Added do loop end `continue`.  **Logic question near line 654: reuse of `iss` might lead to confusion.  The array index `nx` is not used anywhere inside the loop.**   Other files that reuse `iss` are also a bit hard to follow but of course would be more understandable if familiar with the code. |
| `evasec.for` |  Added do loop end `continue`. |
| `execut.for` |  Multiple calls to `outIchk` fixed to correct arguments. |
| `getdiv.for` |  Added do loop end `continue`. |
| `getrep.for` |  Added do loop end `continue`. |
| `getres.for` |  Added do loop end `continue`. |
| `getres2.for` |  Added do loop end `continue`. |
| `makefile` |  Ray had trouble with `OutWW.for` compared to `outww.for` filename.  See comment below for `outWW.for` file. |
| `mdainp.for` |  Split arithmetic if statement into multiple if statements.  Added do loop end `continue`.  |
| `ouflow.for` |  Added do loop end `continue`. |
| `outbal2.for` |  Added do loop end `continue`. |
| `outdivw.for` |  Added do loop end `continue`. |
| `outichk.for` |  Minor reformatting. |
| `outmon.for` |  Added do loop end `continue`. |
| `outwel.for` |  Added do loop end `continue`. |
| `outwelc.for` |  Added do loop end `continue`. |
| `outww.for` |  Looks like Ray had an issue with filename mixed case.  His previous iteration named the file `outWW.for` but I mistakenly namec `OutWW.for` in the makefile.  I updated the makefile to use `outWW.for`.  |
| `resrpl.for` |  Added do loop end `continue`. |
| `riginp.for` |  Added do loop end `continue`. |
| `rtnmax.for` |  Added do loop end `continue`. |
| `rtnmaxe.for` |  Added do loop end `continue`. |
| `rtnsec.for` |  Added do loop end `continue`. |
| `rtnsecr.for` |  Added do loop end `continue`. |
| `rtnsecx.for` |  Added do loop end `continue`. |
| `rtnxcu.for` |  Added do loop end `continue`. |
| `rtnxcup.for` |  Added do loop end `continue`. |
| `somnmy.for` |  Added do loop end `continue`. |
| `virgain.for` |  Added do loop end `continue`. |

**Comment by smalers** [comment]
Trying to compile the 64-bit `gfortran` executable gives the following:

```
make statemod
gfortran -g -cpp -DCOMPILER=1 -fno-align-commons -static -fcheck=bounds -fno-automatic -finit-local-zero -fcheck=do,mem,pointer -I/usr/include -c outWW.for
# Windows...
gfortran -g -cpp -DCOMPILER=1 -fno-align-commons -static -fcheck=bounds -fno-automatic -finit-local-zero -fcheck=do,mem,pointer -I/usr/include -o statemod-16.00.47-gfortran-win-64bit.exe statem.o namext.o parse_gfortran.o getpath.o license.o skipn.o getfn.o setpar.o virgen.o execut.o report.o xdebug.o accdiv.o accou.o adjcase.o average.o bintop.o bomsec.o calldat.o carrpl.o chekav2.o chekav3.o chekava.o chekpor.o chekres.o chekts.o chkavail.o chkavail2.o chkprf.o chkver.o closs.o coeffa.o comment.o count.o datinp.o dattim_gfortran.o daydata.o daydist.o daydivo.o dayest.o daymon.o dayoutr.o dayreso.o dayset.o daywelo.o demand.o demcons.o deplete.o depletep.o directby.o directex.o directfs.o directwr.o divalt.o divcar.o divcarl.o divcarr.o divcar1.o divcar2.o divimpr.o divimpr2.o divmulti.o divmultR.o divres.o divresp2.o divresr.o divrig.o divrigs.o divrpl.o divrplp.o divrplp2.o divrplr.o dnmfso.o dnmfsow.o dnmfso2.o dsamod.o evasec.o FlowRes.o getcall.o getctlc.o getdiv.o getdiv2.o geteomx.o getfile.o getid.o getin.o getipy2.o getipy4.o getpar.o getpln.o getplnr.o getplnw.o getrch.o getrep.o getres.o getres2.o getrig.o getrtn.o getrtnx.o getrtnw.o getrtnw2.o getsta.o getwel.o getwel2.o getver.o gwsub.o ifrdown.o ifrrig.o ifrrigsp.o ifrrigx.o ifrrig2.o interp.o intersup.o JMFlow.o JMStore.o mdainp.o oopbook2.o oopdiv.o oprdown.o oprexp.o oprfind.o oprinout.o oprinp.o ouflow.o outbal2.o outcallr.o outcalls.o outcu.o outdeb.o outdivc.o outdivw.o outgvc.o outichk.o outifr.o OutJM.o outmon.o outopr.o outpln.o outplnmo.o outpltd.o outpltr.o outpltw.o outrch.o outrchr.o outres.o outresc.o outrev.o outrg.o outriv.o outsp.o outspd.o outsys.o outsyt.o outsyta.o outtbl.o outtop.o outwel.o outwelc.o outwelp.o outwr2.o outWW.o outxss.o outxssmo.o planeva.o powrig.o powres.o powresp.o powres2.o powsea.o powseap.o powsear.o putpath.o replace.o repsort.o resrgp.o resrg1.o resrg1p.o resrpl.o return.o return2.o riginp.o rigsor.o rivrtn.o rgrg.o roundof.o rsrspu.o rsrspup.o rtncarry.o rtnmax.o rtnmaxe.o rtnsec.o rtnsecc.o rtnsecp.o rtnsecr.o rtnsecrp.o rtnsecw.o rtnsecwp.o rtnsecx.o rtnxcu.o rtnxcup.o sepsec.o setcarl.o setdem.o seteff.o setgw.o setlimit.o setloss.o setplano.o setqdiv.o setqdivc.o setqdivx.o settc.o sjrip.o skip.o soilcu.o soilin.o soilm.o somnmy.o splatte.o spruse.o stafind.o takout.o takou2.o test.o vircom.o virgain.o virin.o virnod.o virout.o virset.o welaugp.o welprity.o welrech.o welrig3.o welrig3p.o year.o year2.o
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0\libgfortran.a(read.o):(.text$_gfortrani_convert_real+0xb7): relocation truncated to fit: IMAGE_REL_AMD64_REL32 against undefined symbol `strtoflt128'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0\libgfortran.a(read.o):(.text$_gfortrani_convert_infnan+0xb3): relocation truncated to fit: IMAGE_REL_AMD64_REL32 against undefined symbol `strtoflt128'
collect2.exe: error: ld returned 1 exit status
make: *** [makefile:673: statemod] Error 1
```

Doing some research I found the following:

[Windows regression: Undefined symbol strtoflt128 when using MSYS toolchain
#2721](https://github.com/actions/virtual-environments/issues/2721)

However, this results in:

```
gfortran -g -cpp -DCOMPILER=1 -fno-align-commons -static -fcheck=bounds -fno-automatic -finit-local-zero --default-image-base-low -fcheck=do,mem,pointer -I/usr/include -c skipn.for
gfortran.exe: error: unrecognized command-line option '--default-image-base-low'
```

More research found [2021-01-31 - ASLR enabled by default](https://www.msys2.org/news/#2021-01-31-aslr-enabled-by-default).  Maybe the recommended option was  temporary thing that has now been phased out?

I did a clean compile in MSys 32-bit window and was able to create and run a StateMod executable.  So, the issue is related to 64-bit.  I'll keep researching.

**Comment by smalers** [comment]
Reading the above issue resources more closely recommends adding `-Wl,--default-image-base-low`.  This resulted in an executable that runs.  I will clean up the makefile a bit more to gracefully handle 32 and 64 bit and then keep moving.

**Comment by smalers** [comment]
I made some additional changes and am sending all the code to Ray in a zip file for his further consideration.
Recent changes are summarized below.  Keep open until we resolve issues.

```
common.inc - remove isgi
dattim.for - removed isgi
dattim_gfortran.for - removed isgi and fixed so it works
dattim_ms.for - new routine to compute milliseconds from dattim output, will use in performance testing to do time differences
datttim_string.for - new routine to format dattim output as string in ISO 8601 format for logging, will use for permanence testing
execut.for - remove isgi, fix a few more obvious typos, remove some extra spaces after column 72 that are flagged in editor
isleapyear.for - new routine to determine if a year is a leap year, used  by dattim_ms
makefile - add new routines and checks to make sure a person does not mix 32-bit and 64-bit object files
outtop.for - remove isgi
parse_gfortran.for - increase command line max to 512 characters, minor code cleanup
report.for - remove isgi
resrpl.for - fix typo seen in other files "TEMPARARY" - sometimes when I see a typo I search all the files to see if it needs fixed elsewhere
statem.for - remove isgi and test new time routines (will expand use in performance testing later), minor code cleanup such as history print statements that are beyond column 72
```

**Comment by smalers** [comment]
I am closing this issue.  The code for 16.00.47 has been incorporated. Other issues can be opened if problems arise.
